### DISCUSSION TITLE
How to distinguish between single and double-tap on device without central scene?

### DISCUSSION THREAD

**Discussion by nmpu** [zwave-js/zwave-js#6586] [Q&A] [answered]
This discussion is not specific to a particular device, but we can use **14287/ZW4002** as an example. This device does not support central scene. Assume the fan control is already on.

If I tap up, I'll get:

`38-0-currentValue`

That's nice, but it's the same event you'd get any time the device changes state. If I associate group 2 with the hub, I'll also get:

`32-0-event`

which will only occur for a local tap. This is what I want.

However, this is the same game that's used for double-tap. If I associate group 3 with the hub, then I get the same 32-0-event for _either_ single or double-tap. I'd like to be able to tell them apart.

Is there some way to tweak the event's **endpoint** or **value** so that I can distinguish between single and double tap?

**Comment by AlCalzone** [comment]
> That's nice, but it's the same event you'd get any time the device changes state

This is a HA issue. `node-zwave-js` emits `value updated` events, whether the value changed or not. So if you were able to detect that the value was updated without changing, you would be able to detect the tap up in the on state. However, HA ignores value updates without changes.

> Is there some way to tweak the event's endpoint or value

No. We only expose what the device sends. I'm afraid the solution is to only associate group 3 and lobby the HA devs to add some way to trigger on unchanged value updates.

**Comment by kpine** [comment]
Double-taps for this device are already supported in HA the same as all other GE devices are, using 1) [Scene notification events](https://www.home-assistant.io/integrations/zwave_js/#scene-events-value-notification), or 2) [`event` entities](https://www.home-assistant.io/integrations/event/) (Event value).

![image](https://github.com/zwave-js/node-zwave-js/assets/791794/c9ceaab0-a88b-4bc5-a214-512ab0c24a6f)

**Reply by AlCalzone** [reply]
Double taps, yes. Pressing up once when the device is already on doesn't get exposed, although the information is there.

**Reply by kpine** [reply]
Even if HA supported that, you still can't distinguish local tap events from Z-Wave commands.

**Reply by melyux** [reply]
I get physical (local) tap events in Home Assistant now on this same ZW4002 device, and they're distinguishable from Z-Wave commands which don't trigger these events. So that's good, something has changed. But there's also a 3rd event entity, "Event value", which is a mystery to me.

<img width="300" height="200" alt="image" src="https://github.com/user-attachments/assets/f286f4d4-88d8-4535-86d7-592a45400da0" />

**Reply by kpine** [reply]
There are [tons of revisions](https://products.z-wavealliance.org/?_zwa_search=ZW4002&_per_page=36) of the the ZW4002, the one you have supports Central Scene. The version that started this discussion (14287) does not.

The event entity is for these older devices.

**Reply by melyux** [reply]
Do you know what this means for my revision? It still has an event, and all of mine have different timestamps, just not sure what it represents.

**Reply by kpine** [reply]
Well, press the paddles and see what happens?

I think you'll find the "Event value" is not relevant. Maybe it will send an event with a double tap, you'll have to try, but the other event entities already cover that. All of the newer ZW4002 device config files are copied from the old ones, so they enable the functionality when it's not needed.

**Reply by AlCalzone** [reply]
> All of the newer ZW4002 device config files are copied from the old ones, so they enable the functionality when it's not needed.

Do you happen to know how to distinguish them?

**Reply by nmpu** [reply, marked as answer]
I was the one who started this discussion. I was a little green back then. I did eventually find reliable ways to distinguish every possible event on every device I own. I've been running the same scripts for more than a year now. I can detect these _local_ events:

- Double-tap (with native support)
- Triple-tap (with native Central Scene support)
- Tap up (including when already on, which requires deadtime)
- Tap down (including when already off, which requires deadtime)
- Double-tap down when off and double-tap up when on for older NuTone devices. This could be made [more reliable with Al's help](https://github.com/zwave-js/zwave-js/discussions/6643), but I'm just going to retire these devices.

I had to study the typical Z-Wave event sequences and devise logic to differentiate.

All of my dimmers are templated as 'fans' with calibrated (lux meter) 'dim' and 'on/normal' preset values. 'bright' is always 99. I also have presets for dim_save and normal_save which allow the current value to become the new default. I use _tap up when already on_ to cycle normal -> bright -> dim -> normal. _Tap down when already off_ goes straight to dim.

I use the same single-tap logic for low -> medium -> high on fan controllers. On the fans, I also use _double-tap up_ for medium, _triple-tap up_ for high, and _double-tap down_ for low.

I process all the Z-Wave events and generate my own higher-level events which are then indexed into an action table.

Basically, my setup is data driven. I have less than 20 total automations/scripts. They all make extensive use of conditional Jinja expressions rather than sprawling if/then blocks. I suppose it's more like 'ladder logic'.

I've also done some clever stuff with motion devices. When I enter the garage, the motion sensor turns the light on and will automatically turn off after a short period without motion. If I tap up, it forces the light on for 1 hour. When I turn off a motion device, there's a 5-second exit delay before it will retrigger. When the motion timeout approaches, the lights dim for 10 seconds before turning off. If you move, the lights go back to normal. When you approach the toilet, the light comes on. If motion is detected again after 2 minutes, but before 3 minutes, the exhaust fan come on. That's how I guess number 1 vs number 2. :) I use similar logic for the shower.

Before Home Assistant, I had all these ideas and started messing around with SmartThings 'edge' drivers which were in beta. The documentation and support were terrible. I'll never go back. I need to give those hubs away.

I tell my family that it's likely nobody (not Bill Gates, not Mark Zuckerberg, not Linus Sebastian) have this level of control using simple Decora switches. Neither my wife nor young-adult sons have any interest in the mechanics. They don't even know which box is the brain.

I've never played with a ZigBee device. Does ZigBee use a similar event model? Are double- and triple-tap supported? Is it even possible to detect things like _tap up when already on_? What about the cloud-based Wi-Fi switches? It's definitely possible with ESPHome. It would be sweet if all my high-level events came straight from the switch. The switch itself is omniscient. I've considered attempting custom Z-Wave firmware, but I suspect the tools would be expensive and there's that supposed 'certification' thing that Jasco manages to ignore.

@melyux Sorry to drag you through my technical rant. You need to know that anything is possible, but attempting to build scaffolding around the existing entities is going to disappoint. By that point, contextual information has already been lost.  I had another issue with the way [Z-Wave percentages are scaled](https://community.home-assistant.io/t/conversion-between-brightness-pct-1-100-z-wave-1-99-and-brightness-1-255/661523). I handle that myself as well. At some point, you remove the training wheels and give up the silly visual dropdowns. You can then start to create generic scripts that work with multiple devices.

**Comment by kpine** [comment]
If you want to handle value update events outside of HA's entities, you can use the [`zwave_js.value_updated`](https://www.home-assistant.io/integrations/zwave_js/#zwave_jsvalue_updated) platform trigger. You can use the `38-0-currentValue` value as the trigger and it will execute anytime Z-Wave JS sends a value updated event, even if the device is already on.

You will still have the problem of detecting local taps vs. Z-Wave commands. For example, if you turn on the fan via HA, Z-Wave JS will send an optimistic value update, and then it will send another value update for the manual refresh. These will be indistinguishable from a local tap. You would need some additional automation logic to try and detect whether a Z-Wave command was just sent. Maybe the [trigger state context](https://www.home-assistant.io/docs/configuration/state_object#context) has that information.

For HA to emit a state trigger when the entity's current value is set, but doesn't change, would require modifying the integration to enable the [`force_update`](https://developers.home-assistant.io/docs/core/entity#advanced-properties) attribute. This is already done for `sensor` entities. Enabling this for fan/light/switch may receive push back from HA devs though, your best chance would be to submit a PR and have it rejected/accepted, or ask devs on HA Discord regarding the idea and see if anyone from Core responds. Enabling this would possibly break people's existing automations if they aren't using a `from` clause in their trigger, as they may be assuming the current behavior.

**Reply by nmpu** [reply]
@AlCalzone, as @kpine has mentioned, **HA** does expose all the events. The problem is that `37/38-x-currentValue` also appears every time you set a new target value. The sequence looks like this:
```
Value updated: 38-0-targetValue 10 => 47
Value updated: 38-0-currentValue 10 => 47
Value updated: 38-0-currentValue 47 => 47
```
This means it's impossible to distinguish a local tap (without elaborate context tracking). I'm not suggesting this is a bug in either **Z-Wave JS** or **HA**. It's just how things work. Another difficulty is that `37/38-x-currentValue` can arrive multiple times due to imperfect communications.

When `Value notification: 32-x-event` arrives, can **Z-Wave JS** distinguish between **Group 2** and **Group 3** association? If not, could **Z-Wave JS** _snoop_ on a dedicated proxy node used as the target? Just trying to understand what's possible. I see there are now 'virtual' nodes. Would there be some way to associate with a virtual node and have that virtual node signal the controller?

I currently have automations to detect U0, U1, U2, D0, D1, D2 with or without central scene. U0/D0 are local on/off. **HA** can just barely keep up with the burst of events. Unfortunately, the update triggers do not have a unique timestamp so I have to rely on the (shared) automation timestamp and use parallel mode. Queued mode would be more robust. Everything works. I just have to discard some sequences when things become ambiguous. Since these are all local, it's no big deal to just repeat the tap sequence. I currently use `32-x-event` for both single and double-tap. This 'local' marker gives me another data point to key off. The single-tap up/down must agree with the latest `37/38-x-currentValue`. If I could distinguish between **Group 2** and **Group 3**, I could immediately detect U2/D2 without context (or delay).

Thanks to both of you for entertaining this discussion. I know it's _way off_ the beaten path. Obviously, this would all be non-issue if I got rid of anything that's not central scene.

**Reply by AlCalzone** [reply]
> can Z-Wave JS distinguish between Group 2 and Group 3 association

No. We only receive the report, no information which association group sent it. Even if that were possible, most groups have no meaning to Z-Wave JS.

> Would there be some way to associate with a virtual node and have that virtual node signal the controller?

This could theoretically be possible with a controller based on the Bridge Controller Library, with a virtual node that you set as the association destination. However Z-Wave JS doesn't support this currently - and neither does any application. I'm not sure how HA would even model these virtual nodes.

Other than that I could imagine making this work by adding context to the value update - whether the update was expected (optimistic, response to a GET, ...) or unsolicited (sent by the device without any action from the controller). This would require HA to add support for this flag, but could potentially help with similar automations.
